File: long_questions/DSA/Leetcode/golang/22_Topological_Sort/207_Course_Schedule.go
package main

import "fmt"

// 207. Course Schedule
// Time: O(V+E), Space: O(V+E) where V is number of courses, E is prerequisites
func canFinish(numCourses int, prerequisites [][]int) bool {
	// Build adjacency list and in-degree count
	adj := make(map[int][]int)
	inDegree := make([]int, numCourses)
	
	for _, prereq := range prerequisites {
		course := prereq[0]
		prereqCourse := prereq[1]
		
		adj[prereqCourse] = append(adj[prereqCourse], course)
		inDegree[course]++
	}
	
	// Find courses with no prerequisites (in-degree = 0)
	queue := []int{}
	for i := 0; i < numCourses; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}
	
	// Process courses
	processed := 0
	for len(queue) > 0 {
		course := queue[0]
		queue = queue[1:]
		processed++
		
		// Update in-degree for dependent courses
		for _, neighbor := range adj[course] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}
	
	return processed == numCourses
}

// DFS approach to detect cycles
func canFinishDFS(numCourses int, prerequisites [][]int) bool {
	// Build adjacency list
	adj := make(map[int][]int)
	for _, prereq := range prerequisites {
		course := prereq[0]
		prereqCourse := prereq[1]
		adj[prereqCourse] = append(adj[prereqCourse], course)
	}
	
	// States: 0 = unvisited, 1 = visiting, 2 = visited
	state := make([]int, numCourses)
	
	var dfs func(int) bool
	dfs = func(course int) bool {
		if state[course] == 1 {
			return true // Cycle detected
		}
		if state[course] == 2 {
			return false // Already processed
		}
		
		state[course] = 1 // Mark as visiting
		
		for _, neighbor := range adj[course] {
			if dfs(neighbor) {
				return true
			}
		}
		
		state[course] = 2 // Mark as visited
		return false
	}
	
	// Check all courses for cycles
	for i := 0; i < numCourses; i++ {
		if state[i] == 0 {
			if dfs(i) {
				return false // Cycle detected
			}
		}
	}
	
	return true
}

// Alternative approach using Union-Find to detect cycles
func canFinishUnionFind(numCourses int, prerequisites [][]int) bool {
	parent := make([]int, numCourses)
	for i := 0; i < numCourses; i++ {
		parent[i] = i
	}
	
	var find func(int) int
	find = func(x int) int {
		if parent[x] != x {
			parent[x] = find(parent[x])
		}
		return parent[x]
	}
	
	var union func(int, int) bool
	union = func(x, y int) bool {
		rootX := find(x)
		rootY := find(y)
		
		if rootX == rootY {
			return false // Cycle detected
		}
		
		parent[rootY] = rootX
		return true
	}
	
	for _, prereq := range prerequisites {
		course := prereq[0]
		prereqCourse := prereq[1]
		
		if !union(prereqCourse, course) {
			return false // Cycle detected
		}
	}
	
	return true
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Topological Sort for Cycle Detection
- **Graph Representation**: Courses as nodes, prerequisites as directed edges
- **Cycle Detection**: Check if graph has cycles (impossible schedule)
- **Topological Order**: Valid order exists if no cycles
- **Multiple Approaches**: BFS (Kahn's), DFS, Union-Find

## 2. PROBLEM CHARACTERISTICS
- **Directed Graph**: Courses with prerequisite relationships
- **Cycle Detection**: Determine if valid ordering exists
- **Course Dependencies**: Must complete prerequisites before courses
- **Feasibility Check**: Return true if all courses can be completed

## 3. SIMILAR PROBLEMS
- Course Schedule II (LeetCode 210) - Find valid course order
- Minimum Height Trees (LeetCode 310) - Find tree roots
- Alien Dictionary (LeetCode 269) - Determine dictionary order
- Sequence Reconstruction (LeetCode 444) - Reconstruct sequence

## 4. KEY OBSERVATIONS
- **Graph Theory**: Problem reduces to cycle detection in directed graph
- **Topological Sort**: Valid order exists iff no cycles
- **Multiple Solutions**: BFS, DFS, Union-Find all work
- **Edge Cases**: Empty courses, no prerequisites, single course

## 5. VARIATIONS & EXTENSIONS
- **Course Ordering**: Find actual valid order (LeetCode 210)
- **Multiple Prerequisites**: Courses can have multiple requirements
- **Parallel Courses**: Maximum courses per semester
- **Prerequisite Updates**: Dynamic prerequisite changes

## 6. INTERVIEW INSIGHTS
- Always clarify: "Course numbering? Duplicate prerequisites? Self-loops?"
- Edge cases: empty courses, no prerequisites, cycles
- Time complexity: O(V+E) where V=courses, E=prerequisites
- Space complexity: O(V+E) for graph storage
- Key insight: cycle detection in directed graph

## 7. COMMON MISTAKES
- Not handling self-cycles (course depends on itself)
- Wrong graph direction (prereq → course vs course → prereq)
- Not processing all connected components
- Incorrect cycle detection in DFS
- Union-Find implementation errors

## 8. OPTIMIZATION STRATEGIES
- **BFS (Kahn's)**: O(V+E) time, O(V+E) space
- **DFS Cycle Detection**: O(V+E) time, O(V) space
- **Union-Find**: O(V+E) time, O(V) space
- **Early Termination**: Stop when cycle detected

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like course planning with dependencies:**
- You have courses with prerequisite requirements
- Need to determine if all courses can be completed
- If there's a cycle (A→B→A), impossible to complete
- Like a dependency graph where cycles break the system
- Topological sort finds valid ordering or detects cycles

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Number of courses, prerequisite pairs [course, prereq]
2. **Goal**: Determine if all courses can be completed
3. **Constraint**: Must complete prerequisites before courses
4. **Output**: Boolean indicating feasibility

#### Phase 2: Key Insight Recognition
- **"Graph natural fit"** → Courses as nodes, prereqs as edges
- **"Cycle detection"** → Valid schedule exists iff no cycles
- **"Topological sort"** → Standard algorithm for this problem
- **"Multiple approaches"** → BFS, DFS, Union-Find all work

#### Phase 3: Strategy Development
```
Human thought process:
"I need to check if course dependencies have cycles.
This is a classic topological sort problem:

BFS Approach (Kahn's Algorithm):
1. Build graph and count in-degrees
2. Start with courses having 0 in-degree (no prerequisites)
3. Remove them and update in-degrees of dependent courses
4. If we can remove all courses → no cycles
5. If stuck → cycle exists

DFS Approach:
1. Build adjacency list
2. DFS with cycle detection (visiting state)
3. If we encounter a node being visited → cycle
4. If DFS completes without cycles → valid schedule

Union-Find Approach:
1. For each prerequisite [course, prereq]
2. If course and prereq are already connected → cycle
3. Otherwise, union them
4. If no cycles detected → valid schedule"
```

#### Phase 4: Edge Case Handling
- **No courses**: Return true (nothing to complete)
- **No prerequisites**: Return true (all courses independent)
- **Single course**: Return true unless self-dependent
- **Self-cycle**: Course depends on itself → return false

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Courses: 2, Prerequisites: [[1,0]]

Human thinking:
"Build graph:
Course 0: no prerequisites
Course 1: requires course 0

BFS Approach:
1. Count in-degrees: course 0 has 0, course 1 has 1
2. Queue starts with [0] (no prerequisites)
3. Process course 0:
   - Remove from queue
   - Update course 1 in-degree: 1 → 0
   - Add course 1 to queue
4. Process course 1:
   - Remove from queue
   - No dependents to update
5. All courses processed → no cycles ✓

Result: true (can complete both courses)"
```

#### Phase 6: Intuition Validation
- **Why graph works**: Natural representation of dependencies
- **Why cycle detection works**: Cycles make completion impossible
- **Why topological sort works**: Standard algorithm for DAG validation
- **Why multiple approaches work**: Different ways to detect cycles

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just try all orders?"** → Exponential permutations
2. **"Should I use BFS or DFS?"** → Both work, choose based on preference
3. **"What about Union-Find?"** → Works but less intuitive for this problem
4. **"Can I optimize further?"** → O(V+E) is already optimal
5. **"What about course ordering?"** → Different problem (LeetCode 210)

### Real-World Analogy
**Like planning a curriculum with course prerequisites:**
- You have courses that require other courses first
- Need to determine if the curriculum is feasible
- If there's a circular dependency (A needs B, B needs A), impossible
- Build dependency graph and check for cycles
- Valid curriculum has no prerequisite cycles

### Human-Readable Pseudocode
```
function canFinish(numCourses, prerequisites):
    // Build graph and in-degree count
    adj = adjacency list
    inDegree = array of size numCourses
    
    for each [course, prereq] in prerequisites:
        adj[prereq].append(course)
        inDegree[course]++
    
    // Initialize queue with courses having no prerequisites
    queue = []
    for i from 0 to numCourses-1:
        if inDegree[i] == 0:
            queue.append(i)
    
    // Process courses
    processed = 0
    while queue is not empty:
        course = queue.pop_front()
        processed++
        
        for each neighbor in adj[course]:
            inDegree[neighbor]--
            if inDegree[neighbor] == 0:
                queue.append(neighbor)
    
    return processed == numCourses
```

### Execution Visualization

### Example: numCourses = 2, prerequisites = [[1,0]]
```
Graph Construction:
Course 0: []
Course 1: [0]

In-degree calculation:
inDegree[0] = 0 (no prerequisites)
inDegree[1] = 1 (requires course 0)

BFS Processing:
Initial queue: [0] (courses with in-degree 0)

Step 1: Process course 0
- Remove 0 from queue
- processed = 1
- Update neighbors: course 1 in-degree: 1 → 0
- Add course 1 to queue: [1]

Step 2: Process course 1
- Remove 1 from queue
- processed = 2
- No neighbors to update

Final: processed = 2, numCourses = 2 ✓
All courses can be completed!
```

### Key Visualization Points:
- **Graph Building**: Course → dependent courses
- **In-degree Tracking**: Number of prerequisites per course
- **Queue Processing**: Courses with no prerequisites first
- **Cycle Detection**: Stuck queue indicates cycle

### Memory Layout Visualization:
```
Graph State During Processing:
Courses: 2, Prerequisites: [[1,0]]

Adjacency List:
0: [1]  (course 0 enables course 1)
1: []   (course 1 enables nothing)

In-degree Array:
[0, 1]  (course 0: 0 prereqs, course 1: 1 prereq)

Queue Evolution:
Initial: [0]  (courses with 0 in-degree)
After processing 0: [1]  (course 1 now has 0 in-degree)
After processing 1: []  (all courses processed)

Result: processed=2, total=2 ✓ No cycles!
```

### Time Complexity Breakdown:
- **Graph Building**: O(V+E) time, O(V+E) space
- **BFS Processing**: O(V+E) time, O(V) space for queue
- **DFS Processing**: O(V+E) time, O(V) space for recursion
- **Union-Find**: O(V+E) time, O(V) space
- **Total**: O(V+E) time, O(V+E) space

### Alternative Approaches:

#### 1. DFS Cycle Detection (O(V+E) time, O(V) space)
```go
func canFinishDFS(numCourses int, prerequisites [][]int) bool {
    adj := make(map[int][]int)
    for _, prereq := range prerequisites {
        course, prereqCourse := prereq[0], prereq[1]
        adj[prereqCourse] = append(adj[prereqCourse], course)
    }
    
    state := make([]int, numCourses) // 0=unvisited, 1=visiting, 2=visited
    
    var dfs func(int) bool
    dfs = func(course int) bool {
        if state[course] == 1 {
            return true // Cycle detected
        }
        if state[course] == 2 {
            return false // Already processed
        }
        
        state[course] = 1 // Visiting
        for _, neighbor := range adj[course] {
            if dfs(neighbor) {
                return true
            }
        }
        state[course] = 2 // Visited
        return false
    }
    
    for i := 0; i < numCourses; i++ {
        if state[i] == 0 && dfs(i) {
            return false
        }
    }
    return true
}
```
- **Pros**: Intuitive, works for cycle detection
- **Cons**: Recursion depth, more complex than BFS

#### 2. Union-Find Cycle Detection (O(V+E) time, O(V) space)
```go
func canFinishUnionFind(numCourses int, prerequisites [][]int) bool {
    parent := make([]int, numCourses)
    for i := 0; i < numCourses; i++ {
        parent[i] = i
    }
    
    find := func(x int) int {
        if parent[x] != x {
            parent[x] = find(parent[x])
        }
        return parent[x]
    }
    
    union := func(x, y int) bool {
        rootX, rootY := find(x), find(y)
        if rootX == rootY {
            return false // Cycle detected
        }
        parent[rootY] = rootX
        return true
    }
    
    for _, prereq := range prerequisites {
        course, prereqCourse := prereq[0], prereq[1]
        if !union(prereqCourse, course) {
            return false // Cycle detected
        }
    }
    return true
}
```
- **Pros**: Simple implementation, no recursion
- **Cons**: Less intuitive, path compression needed for efficiency

#### 3. Matrix Representation (O(V²) time, O(V²) space)
```go
func canFinishMatrix(numCourses int, prerequisites [][]int) bool {
    // Use adjacency matrix for small number of courses
    graph := make([][]bool, numCourses)
    for i := 0; i < numCourses; i++ {
        graph[i] = make([]bool, numCourses)
    }
    
    for _, prereq := range prerequisites {
        course, prereqCourse := prereq[0], prereq[1]
        graph[prereqCourse][course] = true
    }
    
    // Apply topological sort using matrix
    // ... (implementation details omitted)
}
```
- **Pros**: Simple for small number of courses
- **Cons**: O(V²) space, inefficient for large V

### Extensions for Interviews:
- **Course Ordering**: Find actual valid order (LeetCode 210)
- **Multiple Prerequisites**: Courses can have multiple requirements
- **Parallel Courses**: Maximum courses per semester
- **Prerequisite Updates**: Dynamic prerequisite changes
- **Performance Analysis**: Discuss worst-case scenarios
*/
func main() {
	// Test cases
	testCases := []struct {
		numCourses  int
		prereqs     [][]int
		description string
	}{
		{2, [][]int{{1, 0}}, "Simple linear dependency"},
		{2, [][]int{{1, 0}, {0, 1}}, "Direct cycle"},
		{4, [][]int{{1, 0}, {2, 0}, {3, 1}, {3, 2}}, "Complex DAG"},
		{1, [][]int{}, "Single course no prerequisites"},
		{3, [][]int{{0, 1}, {0, 2}, {1, 2}}, "Multiple dependencies"},
		{3, [][]int{{0, 1}, {1, 2}, {2, 0}}, "3-course cycle"},
		{5, [][]int{{1, 0}, {2, 0}, {3, 1}, {4, 2}, {4, 3}}, "Complex structure"},
		{0, [][]int{}, "No courses"},
		{4, [][]int{{2, 0}, {1, 0}, {3, 1}, {3, 2}, {1, 3}}, "Complex with cycle"},
		{4, [][]int{{1, 0}, {2, 1}, {3, 2}}, "Linear chain"},
	}
	
	for i, tc := range testCases {
		fmt.Printf("Test Case %d: %s\n", i+1, tc.description)
		fmt.Printf("  Courses: %d, Prerequisites: %v\n", tc.numCourses, tc.prereqs)
		
		result1 := canFinish(tc.numCourses, tc.prereqs)
		result2 := canFinishDFS(tc.numCourses, tc.prereqs)
		result3 := canFinishUnionFind(tc.numCourses, tc.prereqs)
		
		fmt.Printf("  BFS (Kahn's): %t\n", result1)
		fmt.Printf("  DFS: %t\n", result2)
		fmt.Printf("  Union-Find: %t\n\n", result3)
	}
}

File: long_questions/DSA/Leetcode/java/22_Topological_Sort/207_Course_Schedule.java
import java.util.*;

public class CourseSchedule {
    
    // 207. Course Schedule
    // Time: O(V+E), Space: O(V+E) where V is number of courses, E is prerequisites
    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        // Build adjacency list and in-degree count
        Map<Integer, List<Integer>> adj = new HashMap<>();
        int[] inDegree = new int[numCourses];
        
        for (int[] prereq : prerequisites) {
            int course = prereq[0];
            int prereqCourse = prereq[1];
            
            adj.computeIfAbsent(prereqCourse, k -> new ArrayList<>()).add(course);
            inDegree[course]++;
        }
        
        // Find courses with no prerequisites (in-degree = 0)
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }
        
        // Process courses
        int processed = 0;
        while (!queue.isEmpty()) {
            int course = queue.poll();
            processed++;
            
            // Update in-degree for dependent courses
            for (int neighbor : adj.getOrDefault(course, new ArrayList<>())) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        
        return processed == numCourses;
    }

    // DFS approach to detect cycles
    public static boolean canFinishDFS(int numCourses, int[][] prerequisites) {
        // Build adjacency list
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int[] prereq : prerequisites) {
            int course = prereq[0];
            int prereqCourse = prereq[1];
            adj.computeIfAbsent(prereqCourse, k -> new ArrayList<>()).add(course);
        }
        
        // States: 0 = unvisited, 1 = visiting, 2 = visited
        int[] state = new int[numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            if (state[i] == 0 && hasCycleDFS(i, adj, state)) {
                return false; // Cycle detected
            }
        }
        
        return true;
    }
    
    private static boolean hasCycleDFS(int course, Map<Integer, List<Integer>> adj, int[] state) {
        if (state[course] == 1) {
            return true; // Cycle detected
        }
        if (state[course] == 2) {
            return false; // Already processed
        }
        
        state[course] = 1; // Mark as visiting
        
        for (int neighbor : adj.getOrDefault(course, new ArrayList<>())) {
            if (hasCycleDFS(neighbor, adj, state)) {
                return true;
            }
        }
        
        state[course] = 2; // Mark as visited
        return false;
    }

    // Alternative approach using Union-Find to detect cycles
    public static boolean canFinishUnionFind(int numCourses, int[][] prerequisites) {
        int[] parent = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            parent[i] = i;
        }
        
        for (int[] prereq : prerequisites) {
            int course = prereq[0];
            int prereqCourse = prereq[1];
            
            if (!union(prereqCourse, course, parent)) {
                return false; // Cycle detected
            }
        }
        
        return true;
    }
    
    private static int find(int x, int[] parent) {
        if (parent[x] != x) {
            parent[x] = find(parent[x], parent);
        }
        return parent[x];
    }
    
    private static boolean union(int x, int y, int[] parent) {
        int rootX = find(x, parent);
        int rootY = find(y, parent);
        
        if (rootX == rootY) {
            return false; // Cycle detected
        }
        
        parent[rootY] = rootX;
        return true;
    }

    public static void main(String[] args) {
        // Test cases
        Object[][] testCases = {
            {2, new int[][]{{1, 0}}, "Simple linear dependency"},
            {2, new int[][]{{1, 0}, {0, 1}}, "Direct cycle"},
            {4, new int[][]{{1, 0}, {2, 0}, {3, 1}, {3, 2}}, "Complex DAG"},
            {1, new int[][]{}, "Single course no prerequisites"},
            {3, new int[][]{{0, 1}, {0, 2}, {1, 2}}, "Multiple dependencies"},
            {3, new int[][]{{0, 1}, {1, 2}, {2, 0}}, "3-course cycle"},
            {5, new int[][]{{1, 0}, {2, 0}, {3, 1}, {4, 2}, {4, 3}}, "Complex structure"},
            {0, new int[][]{}, "No courses"},
            {4, new int[][]{{2, 0}, {1, 0}, {3, 1}, {3, 2}, {1, 3}}, "Complex with cycle"},
            {4, new int[][]{{1, 0}, {2, 1}, {3, 2}}, "Linear chain"}
        };
        
        for (int i = 0; i < testCases.length; i++) {
            int numCourses = (int) testCases[i][0];
            int[][] prereqs = (int[][]) testCases[i][1];
            String description = (String) testCases[i][2];
            
            System.out.printf("Test Case %d: %s\n", i + 1, description);
            System.out.printf("  Courses: %d, Prerequisites: %s\n", numCourses, Arrays.deepToString(prereqs));
            
            boolean result1 = canFinish(numCourses, prereqs);
            boolean result2 = canFinishDFS(numCourses, prereqs);
            boolean result3 = canFinishUnionFind(numCourses, prereqs);
            
            System.out.printf("  BFS (Kahn's): %b\n", result1);
            System.out.printf("  DFS: %b\n", result2);
            System.out.printf("  Union-Find: %b\n\n", result3);
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Topological Sort
- **Course Dependencies**: Directed acyclic graph of prerequisites
- **Cycle Detection**: Check if course schedule is possible
- **Topological Order**: Linear ordering respecting dependencies
- **Graph Processing**: BFS with in-degree counting

## 2. PROBLEM CHARACTERISTICS
- **Directed Graph**: Courses as nodes, prerequisites as edges
- **Cycle Detection**: Cycles make scheduling impossible
- **Topological Sort**: Find valid course ordering
- **Prerequisite Chain**: Course A must be taken before B if A→B

## 3. SIMILAR PROBLEMS
- Course Schedule II (find order)
- Alien Dictionary
- Minimum Height of Buildings
- Build Order from Projects

## 4. KEY OBSERVATIONS
- Course schedule possible iff graph has no cycles
- Kahn's algorithm uses in-degree counting
- DFS can detect cycles using visitation states
- Union-Find can detect cycles during edge processing
- Time complexity: O(V+E) where V=courses, E=prerequisites

## 5. VARIATIONS & EXTENSIONS
- Find actual course ordering
- Multiple semesters/parallel courses
- Weighted prerequisites
- Different constraint types

## 6. INTERVIEW INSIGHTS
- Clarify: "Are prerequisites guaranteed to be valid?"
- Edge cases: no courses, single course, self-loops
- Time complexity: O(V+E) vs O(V×E) naive
- Space complexity: O(V+E) for graph storage

## 7. COMMON MISTAKES
- Not handling self-loops properly
- Confusing course and prerequisite order
- Missing cycle detection
- Incorrect in-degree initialization

## 8. OPTIMIZATION STRATEGIES
- Use adjacency list for sparse graphs
- Early cycle detection in Union-Find
- In-place processing when possible
- Choose appropriate algorithm based on constraints

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like planning course registration:**
- You have courses with prerequisites (must take A before B)
- This creates a dependency graph (directed edges)
- You can only take a course if all prerequisites are completed
- If there's a cycle (A→B→C→A), no valid schedule exists
- Need to check if the dependency graph is acyclic

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Number of courses and prerequisite pairs
2. **Goal**: Can you finish all courses?
3. **Output**: Boolean indicating if schedule is possible

#### Phase 2: Key Insight Recognition
- **"What makes schedule impossible?"** → Cycles in dependencies
- **"How to detect cycles?"** → Topological sort or DFS
- **"What's the graph structure?"** → Courses as nodes, prerequisites as edges
- **"Why topological sort?"** → Linear ordering respects dependencies

#### Phase 3: Strategy Development
```
Human thought process:
"I'll check for cycles:
1. Build adjacency list and in-degree counts
2. Find courses with no prerequisites (in-degree = 0)
3. Process these courses, reduce in-degree of dependents
4. If we can process all courses, no cycles exist
5. If some courses remain unprocessed, there's a cycle"
```

#### Phase 4: Edge Case Handling
- **No courses**: Return true (nothing to schedule)
- **Single course**: Return true unless it has self-prerequisite
- **Self-loop**: Course requires itself → return false
- **Disconnected components**: Handle independently

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Courses: 4, Prerequisites: [[1,0], [2,0], [3,1], [3,2]]

Human thinking:
"Let's build the dependency graph:
Course 0 ← (no prerequisites)
Course 1 ← Course 0
Course 2 ← Course 0
Course 3 ← Courses 1 and 2

In-degree counts:
Course 0: 0 (no prerequisites)
Course 1: 1 (needs course 0)
Course 2: 1 (needs course 0)
Course 3: 2 (needs courses 1 and 2)

Process courses with in-degree = 0:
Queue: [0], processed = 0

Process course 0:
- Remove from queue, processed = 1
- Update dependents:
  - Course 1: in-degree becomes 0 → add to queue
  - Course 2: in-degree becomes 0 → add to queue
Queue: [1, 2]

Process course 1:
- Remove from queue, processed = 2
- Update dependent:
  - Course 3: in-degree becomes 1 (was 2)
Queue: [2]

Process course 2:
- Remove from queue, processed = 3
- Update dependent:
  - Course 3: in-degree becomes 0 → add to queue
Queue: [3]

Process course 3:
- Remove from queue, processed = 4
- No dependents to update
Queue: []

All 4 courses processed → No cycles ✓"
```

#### Phase 6: Intuition Validation
- **Why it works**: Topological sort exists iff graph is acyclic
- **Why it's efficient**: Each edge processed once
- **Why it's correct**: Kahn's algorithm is proven cycle detection

### Common Human Pitfalls & How to Avoid Them
1. **"Why not try all orders?"** → Exponential O(V!)
2. **"What about DFS?"** → Works but more complex
3. **"How to handle multiple prerequisites?"** → Sum in-degree counts
4. **"What about disconnected components?"** → Process independently

### Real-World Analogy
**Like planning a multi-step project:**
- You have tasks (courses) that depend on each other
- Task B cannot start until Task A is complete
- Some tasks have multiple prerequisites
- You want to know if the project can be completed
- If there's a circular dependency (A→B→C→A), project is impossible
- Topological sort gives you a valid execution order

### Human-Readable Pseudocode
```
function canFinish(numCourses, prerequisites):
    // Build graph
    adj = adjacency list
    inDegree = array of size numCourses, initialized to 0
    
    for [course, prereq] in prerequisites:
        adj[prereq].add(course)
        inDegree[course]++
    
    // Find courses with no prerequisites
    queue = all courses with inDegree[i] == 0
    processed = 0
    
    while queue is not empty:
        course = queue.dequeue()
        processed++
        
        for dependent in adj[course]:
            inDegree[dependent]--
            if inDegree[dependent] == 0:
                queue.enqueue(dependent)
    
    return processed == numCourses
```

### Execution Visualization

### Example: 4 courses, [[1,0], [2,0], [3,1], [3,2]]
```
Dependency Graph:
0 ← (no prerequisites)
1 ← 0
2 ← 0
3 ← 1, 2

In-degree Evolution:
Initial: [0, 1, 1, 2]
After processing 0: [0, 0, 0, 1]
After processing 1: [0, 0, 0, 1]
After processing 2: [0, 0, 0, 0]
After processing 3: [0, 0, 0, 0]

Queue Processing:
Start: [0] → Process 0 → Queue: [1,2]
Process 1 → Queue: [2] → Process 2 → Queue: [3]
Process 3 → Queue: [] → Done

Result: All 4 courses processed ✓

Visualization:
0 → 1,2 (depend on 0)
   ↓
   3 (depends on 1,2)
```

### Key Visualization Points:
- **Graph building** from prerequisites
- **In-degree counting** for each course
- **Queue processing** of courses with no unmet prerequisites
- **Cycle detection** when some courses remain unprocessed
- **Topological order** emerges from processing sequence

### Memory Layout Visualization:
```
Graph Representation:
Adjacency List:
0: [1, 2]
1: [3]
2: [3]
3: []

In-degree Array:
Index: 0  1  2  3
Value: 0  1  1  2

Processing Evolution:
Step 1: Queue=[0], Processed=0
Step 2: Queue=[1,2], Processed=1
Step 3: Queue=[2], Processed=2
Step 4: Queue=[3], Processed=3
Step 5: Queue=[], Processed=4

Final: Processed=4, Total=4 → Success ✓
```

### Time Complexity Breakdown:
- **Graph Building**: O(E) where E is number of prerequisites
- **In-degree Calculation**: O(E) during graph building
- **Queue Processing**: O(V+E) where V is number of courses
- **Total**: O(V+E) time, O(V+E) space
- **Optimal**: Linear time for this problem
- **vs DFS**: O(V+E) time but recursion overhead
*/
